File: test/test_get_clip_volume_data.cc
/**
 * (c) Chad Walker, Chris Kirmse
 */

#include <stdio.h>

#include <string>

#include "../utils.h"
#include "../video_reader.h"

#include "file_io_group.h"

void logProgress(int step, int total) {
    printf("progress %i/%i\n", step, total);
}

int main(int argc, char **argv) {
    if (argc < 4) {
        printf("Need filename to read, start_time, and end_time\n");
        return 1;
    }

    std::string source_pathname = argv[1];

    double start_time = std::stod(argv[2]);
    double end_time = std::stod(argv[3]);

    Avalanche::setDefaultLogFunc();

    printf("lavf version %s\n", Avalanche::getAvFormatVersionString().c_str());

    FileIoGroup file_io_group;

    Avalanche::VideoReader video_reader;

    if (!video_reader.init(&file_io_group, source_pathname)) {
        printf("video reader init failed\n");
        return 1;
    }

    if (!video_reader.verifyHasAudioStream()) {
        printf("video has no audio stream\n");
        return 1;
    }

    Avalanche::GetVolumeDataResult get_volume_data_result;
    if (!video_reader.getClipVolumeData(start_time, end_time, get_volume_data_result, logProgress)) {
        printf("failed to get clip volume data\n");
        return 1;
    }
    printf("mean_volume %f dB max_volume %f dB\n", get_volume_data_result.mean_volume, get_volume_data_result.max_volume);

    return 0;
}

File: nodejs_wrapper/test_get_clip_volume_data.js
/**
 * (c) Chad Walker, Chris Kirmse
 */

import log from '../log.js';

import Avalanche from '../avalanche.js';
import ResourceIo from '../resource_io.js';

const main = async function () {
  if (process.argv.length !== 5) {
    log.info('usage: test_get_clip_volume_data.js <source_filename> <start_time> <end_time>');
    return;
  }

  log.info('lavf version', Avalanche.getAvFormatVersionString());

  const sourceUri = process.argv[2];
  const resourceIo = new ResourceIo(sourceUri);
  let clipVolumeData;
  try {
    const videoReader = Avalanche.createVideoReader();
    await videoReader.init(resourceIo);
    clipVolumeData = await videoReader.getClipVolumeData(
      parseFloat(process.argv[3]),
      parseFloat(process.argv[4]),
      (step, total) => {
        log.info('progress', step, total);
      },
    );
  } catch (err) {
    log.info('failed to get clip volume data', err);
    return;
  } finally {
    Avalanche.destroy();
  }
  log.info('result is', clipVolumeData);
};

main();
